### ISSUE TITLE
wayland fedora 36 

### ISSUE THREAD

**Issue by Akimm113** [veyon/veyon#860]
Doesn't show image on wayland, works on xorg after service restart. Fedora 36

**Comment by barak** [comment]
The transition from X11 to Wayland is pretty far along: Wayland is the default for Redhat, Debian, and Ubuntu. It's still possible to run X11 sessions for many desktop environments (like Gnome) but X11 will start withering away pretty soon.

**Comment by tobydox** [comment]
I'm aware of that. Unfortunately Wayland has no mechanism for capturing the screen without explicit consent of the user (through the Desktop Portal interface). While it's fully understandable from a security point of view, it effectively prevents monitoring applications such as Veyon to work properly in that case since a student explicitly has to grant the access every time. Teachers would always have to rely on the cooperation on the students. So unless there's another technical approach/interface we have to tell Veyon users to continue using X11 :-(

**Comment by barak** [comment]
This seems like something that the Wayland people might want to know about? It seems straightforward in principle to allow some sort of extra authentication mechanism for root processes or something like that. Or maybe the desktop portal interface can be hotwired to automatically and silently consent when appropriate.

**Comment by tobydox** [comment]
Yes, a technical solution granting special apps immediate and permanent access to user sessions would be optimal. Any support in getting such an extension done is welcome.

I already started both a Pipewire as well as WayVNC-based implementation 2 years ago (https://github.com/veyon/veyon/tree/pipewire) but did not continue the work on it due to the mentioned challenges.

**Comment by Conan-Kudo** [comment]
This is something that should be requested in the xdg-desktop-portal repo: https://github.com/flatpak/xdg-desktop-portal

That way, a solution can be designed that is desktop agnostic that can be deployed for managed machines like ones deployed by schools.

**Comment by Conan-Kudo** [comment]
I've created an issue about this, please add (useful!) commentary to provide context for the developers there: https://github.com/flatpak/xdg-desktop-portal/issues/1105

**Comment by idoric** [comment]
@tobydox On the issue created by @Conan-Kudo (see previous comment), there is a question which deserves your clarification since Veyon is directly concerned. The question is: "can you expand exactly what scenarios Veyon needs pre-authorization?"

**Comment by mercifulboss** [comment]
> I'm aware of that. Unfortunately Wayland has no mechanism for capturing the screen without explicit consent of the user (through the Desktop Portal interface). While it's fully understandable from a security point of view, it effectively prevents monitoring applications such as Veyon to work properly in that case since a student explicitly has to grant the access every time. Teachers would always have to rely on the cooperation on the students. So unless there's another technical approach/interface we have to tell Veyon users to continue using X11 :-(

What about using xwaylandvideobridge? It allows screensharing on basically every app with no problems. This bypasses all the problems you mentioned and works on every browser, zoom, teams, discord, etc.

https://blog.davidedmundson.co.uk/blog/xwaylandvideobridge/

**Comment by idoric** [comment]
What the Epoptes project plans to do on this subject can also serve as a source of inspiration: https://github.com/epoptes/epoptes/issues/184.

**Comment by idoric** [comment]
> I'm aware of that. Unfortunately Wayland has no mechanism for capturing the screen without explicit consent of the user (through the Desktop Portal interface). While it's fully understandable from a security point of view, it effectively prevents monitoring applications such as Veyon to work properly in that case since a student explicitly has to grant the access every time. Teachers would always have to rely on the cooperation on the students. So unless there's another technical approach/interface we have to tell Veyon users to continue using X11 :-(
— @tobydox 

It’s possible to automatically run `flatpak permission-set`¹ when all users log in by placing a .desktop file in` /etc/xdg/autostart/`². The idea is not mine, I found it [here](https://github.com/flatpak/xdg-desktop-portal/issues/471#issue-599474847).

The solution is not ideal because the user can revoke the permission, but:
-  if the students are young the problem should remain theoretical, and in any case it will be seen immediately and the student can be sanctioned (personally I have students under the age of fifteen).
- this solution would allow you to move forward, to show that the problem is not theoretical but that there is a concrete software and not a vaporware, which requires forcing authorizations from computer users under Wayland/Portal.

> I've created an issue about this, please add (useful!) commentary to provide context for the developers there: https://github.com/flatpak/xdg-desktop-portal/issues/1105
— @Conan-Kudo 

I saw that your issue was closed as not planned, hence my suggestion above.

¹ https://docs.flatpak.org/en/latest/flatpak-command-reference.html#flatpak-permission-set (unfortunately this deserves many more concrete examples)
² https://specifications.freedesktop.org/autostart-spec/autostart-spec-latest.html

**Comment by barak** [comment]
There are scenarios where it is reasonable for students to choose whether to give consent. (Projecting some student's screen to the whole class.) There are scenarios where students must give consent, but they should do it explicitly and get yelled at if they don't. (Taking a test and the invigilators need to monitor them, or a classroom where the teacher wants to see that students are goofing off on their computers.) And there are scenarios where consent should be turned on automatically and cannot be disabled. (A compute cluster at a spy agency, say, or a similar environment where confidential information is being examined and it should be impossible to disable facilities used to monitor to ensure that all access is appropriate.) Maybe there should be a little warning icon in the corner of the screen "external screen monitor enabled".

**Comment by jkeltz** [comment]
Although not directly related to Fedora 36, I am upgrading from Rocky Enterprise Linux 8 (RHEL8) to Rocky Enterprise Linux 10 (RHEL10) at the end of the summer term in our educational environment. We use Veyon to allow a faculty member to monitor student screens while they are writing tests and exams in the lab.  On RHEL10, I only have the choice of Wayland - no XOrg anymore.  Has Veyon still not come up with any solution for this problem?  Surely there has to be some way to make this work.

**Comment by tobydox** [comment]
Thanks for all the input. We'll come up with an initial implementation this year which will allow allow accessing screens under Wayland but unconditional access permissions remain an issue. It can be solved for KDE (by using flatpak pre-authorization) but with GNOME still a solution needs to be found.

**Comment by jkeltz** [comment]
In my case, since I'm using RHEL, GNOME is really the only option.  That being said, if students have to click an "approve" button to approve screen sharing, while not ideal, it's not the end of the world.  No screen sharing active, no grade.  Problem "solved".

**Comment by Conan-Kudo** [comment]
> In my case, since I'm using RHEL, GNOME is really the only option. That being said, if students have to click an "approve" button to approve screen sharing, while not ideal, it's not the end of the world. No screen sharing active, no grade. Problem "solved".

You can make your RHEL installations with KDE Plasma instead, since KDE Plasma is available in EPEL. Pretty much all CentOS Stream derivatives except Red Hat Enterprise Linux offer install images with KDE Plasma as well. And for Red Hat Enterprise Linux, you can manually set it up or make your own custom deploy images with KDE Plasma.

**Comment by tobydox** [comment]
As of Veyon 4.11 we have experimental Wayland support.
